### PR TITLE
(5.5) Handle NVMe devices during upgrade

### PR DESCRIPTION
**Pull request by r0mant**
## Description

This PR addresses https://github.com/gravitational/gravity/issues/1585. It adds a (hidden) `--docker-device`  flag to the upgrade command in order to be able to pass a new name for the Docker device when old (5.0) clusters that use devicemapper on NVMe disks are being upgraded to 5.5. NVMe device names can change after reboot so this would allow a customer to provide a symlink to a correct block device. See the linked ticket for more information.

The provided Docker device name (or symlink) is assumed to be the same on all cluster nodes. The customer who this is relevant for confirmed that this is a fair assumption.

## Type of change

* New feature (non-breaking change which adds functionality)
* This change has a user-facing impact

## Linked tickets and other PRs

* Closes https://github.com/gravitational/gravity/issues/1585.

## TODOs

- [x] Self-review the change
- [x] Perform manual testing
- [x] Address review feedback

## Implementation

I had to make sure to extend upgrade pre-flight checks to test the provided Docker device to make sure it exists on all nodes so it is threaded from the flag to the phase data and eventually the checker.

## Testing done

This will be a long one... Tested on the AWS EC2 node with NVMe disks to validate the exact behavior the customers would experience.

#### 0. Have a node with NVMe disks

```
[centos@ip-172-31-41-223 gravity-5.0.35]$ lsblk
NAME                                                                                         MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
nvme0n1                                                                                      259:3    0   50G  0 disk
└─nvme0n1p1                                                                                  259:4    0   50G  0 part /
nvme1n1                                                                                      259:0    0   20G  0 disk
nvme2n1                                                                                      259:1    0   15G  0 disk
nvme3n1                                                                                      259:2    0  100G  0 disk
nvme4n1                                                                                      259:5    0   25G  0 disk
```

#### 1. Install an old 5.0 version with devicemapper

```
[centos@ip-172-31-41-223 gravity-5.0.35]$ sudo ./gravity install --advertise-addr=172.31.41.223 --storage-driver=devicemapper --docker-device=/dev/nvme4n1 --cloud-provider=generic
```

#### 2. Verify device used for Docker data

```
[centos@ip-172-31-41-223 gravity-5.0.35]$ lsblk
NAME                                                                                         MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
nvme0n1                                                                                      259:3    0   50G  0 disk
└─nvme0n1p1                                                                                  259:4    0   50G  0 part /
nvme1n1                                                                                      259:0    0   20G  0 disk
nvme2n1                                                                                      259:1    0   15G  0 disk
nvme3n1                                                                                      259:2    0  100G  0 disk
nvme4n1                                                                                      259:5    0   25G  0 disk
├─docker-thinpool_tmeta                                                                      253:0    0  252M  0 lvm
│ └─docker-thinpool                                                                          253:2    0 23.8G  0 lvm
│   ├─docker-259:4-80049534-3d464c135c3ac17b05028589253d3aebaee974a683caa184d8ad63ac83560226 253:3    0   10G  0 dm
```

#### 4. Verify device name has changed

```
[centos@ip-172-31-41-223 ~]$ lsblk
NAME                                                                                              MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
nvme0n1                                                                                           259:4    0   50G  0 disk
└─nvme0n1p1                                                                                       259:5    0   50G  0 part /
nvme1n1                                                                                           259:2    0   20G  0 disk
nvme2n1                                                                                           259:1    0   15G  0 disk
nvme3n1                                                                                           259:0    0   25G  0 disk
├─docker-thinpool_tmeta                                                                           253:0    0  252M  0 lvm
│ └─docker-thinpool                                                                               253:2    0 23.8G  0 lvm
│   ├─docker-259:5-80049534-65927ff7beed55de32a326e4fa6cf0fa455356a4e8a257d0945b5c139ed32c6c      253:3    0   10G  0 dm
```

#### 5. Make a symlink for the new device

```
[centos@ip-172-31-41-223 ~]$ sudo ln -s /dev/nvme3n1 /dev/xvdb
[centos@ip-172-31-41-223 ~]$ ls -l /dev | grep nvme
...
lrwxrwxrwx. 1 root root          12 Jun 10 02:59 xvdb -> /dev/nvme3n1
```

#### 6. Try to upgrade to this branch and get an error

```
[centos@ip-172-31-41-223 gravity-5.5]$ sudo ./upgrade
Wed Jun 10 03:16:52 UTC	Importing cluster image telekube v5.5.48-dev.8
Wed Jun 10 03:17:42 UTC	Synchronizing application with Docker registry 172.31.41.223:5000
Wed Jun 10 03:18:15 UTC	Cluster image has been uploaded
Wed Jun 10 03:18:15 UTC	Upgrading cluster from 5.0.35 to 5.5.48-dev.8
[ERROR]:
The following cluster nodes use NVMe block devices for their Docker data:

Hostname                                        IP                Docker Device
--------                                        --                -------------
ip-172-31-41-223.us-west-2.compute.internal     172.31.41.223     /dev/nvme4n1

NVMe drive labels may change after reboot which will lead to the device names
...
```

#### 7. Launch upgrade with `--docker-device` flag

```
[centos@ip-172-31-41-223 gravity-5.5]$ sudo ./upgrade --docker-device=/dev/xvdb
Wed Jun 10 03:20:05 UTC	Importing cluster image telekube v5.5.48-dev.8
Wed Jun 10 03:20:53 UTC	Synchronizing application with Docker registry 172.31.41.223:5000
Wed Jun 10 03:21:07 UTC	Cluster image has been uploaded
Wed Jun 10 03:21:08 UTC	Upgrading cluster from 5.0.35 to 5.5.48-dev.8
Wed Jun 10 03:21:09 UTC	Deploying agents on cluster nodes
Wed Jun 10 03:21:11 UTC	Deployed agent on ip-172-31-41-223.us-west-2.compute.internal (172.31.41.223)
```

#### 8. Make sure the operation completes successfully

```
[centos@ip-172-31-41-223 gravity-5.5]$ sudo gravity status
Cluster name:		kindvarahamihira8699
Cluster status:		active
Application:		telekube, version 5.5.48-dev.8
Gravity version:	5.5.48-dev.8 (client) / 5.5.48-dev.8 (server)
```

#### 9. Make sure the proper device has been repurposed for overlay

```
[centos@ip-172-31-41-223 gravity-5.5]$ lsblk
NAME        MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
nvme0n1     259:4    0   50G  0 disk
└─nvme0n1p1 259:5    0   50G  0 part /
nvme1n1     259:2    0   20G  0 disk
nvme2n1     259:1    0   15G  0 disk
nvme3n1     259:0    0   25G  0 disk /var/lib/gravity/planet/docker
nvme4n1     259:3    0  100G  0 disk
```

#### 11. Make sure the mount still points to the proper device even if name has changed

```
[centos@ip-172-31-41-223 ~]$ lsblk
NAME        MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
nvme0n1     259:3    0   50G  0 disk
└─nvme0n1p1 259:4    0   50G  0 part /
nvme1n1     259:0    0   20G  0 disk
nvme2n1     259:1    0   25G  0 disk /var/lib/gravity/planet/docker
nvme3n1     259:2    0  100G  0 disk
```

### CODE REVIEW

**Comment by knisbet** [comment]
This is only specific to 5.5 right? Not a permanent change to the code base, just a bridge to 5.5+ / overlay2?

**Comment by r0mant** [comment]
@knisbet Yes, only for 5.5.

File: tool/gravity/cli/register.go [hunk 1]
```diff
@@ -174,6 +174,8 @@ func RegisterCommands(app *kingpin.Application) *Application {
 	g.UpdateTriggerCmd.App = g.UpdateTriggerCmd.Arg("app", "Application version to update to, in the 'name:version' or 'name' (for latest version) format. If unspecified, currently installed application is updated").String()
 	g.UpdateTriggerCmd.Manual = g.UpdateTriggerCmd.Flag("manual", "Manual operation. Do not trigger automatic update").Short('m').Bool()
 	g.UpdateTriggerCmd.SkipVersionCheck = g.UpdateTriggerCmd.Flag("skip-version-check", "Bypass version compatibility check").Hidden().Bool()
+	g.UpdateTriggerCmd.Force = g.UpdateTriggerCmd.Flag("force", "Bypass some pre-upgrade checks").Hidden().Bool()
```
Comment (a-palchikov) [inline tool/gravity/cli/register.go:177]: nit: can't we be specific in this description? `Bypass docker device validation.`?

**Review by a-palchikov — APPROVED** [review]

File: tool/gravity/cli/register.go (hunk 1 above)
Comment (r0mant) [reply tool/gravity/cli/register.go]: This flag has now broader meaning allowing to bypass active warnings, I've updated description accordingly.

**Pull request merged**